pkg/updater/repository/repository.go

```diff
@@ -241,26 +311,134 @@ func movePackageFromSource(packageName string, rootPath string, sourcePath strin
-		isExperiment := r.experiment.Exists() && r.experiment.Target() == file.Name()
+
+	// For each version, get the running PIDs. These PIDs are written by the injector.
+	// The injector is ran directly with the service as it's a LD_PRELOAD, and has access
+	// to the PIDs.
+	for _, version := range versions {
+		isLink := version.Name() == stableVersionLink || version.Name() == experimentVersionLink
+		isStable := r.stable.Exists() && r.stable.Target() == version.Name()
+		isExperiment := r.experiment.Exists() && r.experiment.Target() == version.Name()
 		if isLink || isStable || isExperiment {
 			continue
 		}
-		err := os.RemoveAll(filepath.Join(r.rootPath, file.Name()))
+
+		if !version.IsDir() {
+			continue
+		}
+
+		versionInUse, err := packageVersionInUse(filepath.Join(r.runPath, version.Name()))
+		if err != nil {
+			log.Errorf("could not check if package version is in use: %v", err)
+			continue
+		}
+
+		// If no PIDs are running, remove the version
+		if !versionInUse {
+			log.Debugf("no running PIDs for package %s version %s, removing package", r.rootPath, version.Name())
+			if err := os.RemoveAll(filepath.Join(r.rootPath, version.Name())); err != nil {
+				log.Errorf("could not remove package directory for version %s: %v", version, err)
+			}
+			if err := os.RemoveAll(filepath.Join(r.runPath, version.Name())); err != nil {
+				log.Errorf("could not remove run directory for version %s: %v", version, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// packageVersionInUse checks if the given package version is in use
+func packageVersionInUse(runPath string) (bool, error) {
+	runningPIDs := 0
+
+	pids, err := os.ReadDir(runPath)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Debugf("package run directory does not exist, no running PIDs")
+			return false, nil
+		}
+		return false, fmt.Errorf("could not read run directory: %w", err)
+	}
+
+	// For each PID, check if it's running
+	for _, rawPID := range pids {
+		pid, err := strconv.ParseInt(rawPID.Name(), 10, 64)
+		if err != nil {
+			log.Errorf("could not parse PID: %v", err)
+			continue
+		}
+		processExists, err := process.PidExists(int32(pid))
 		if err != nil {
-			return fmt.Errorf("could not remove file: %w", err)
+			log.Errorf("could not find process with PID %d: %v", pid, err)
+		} else if processExists {
+			runningPIDs++
+		} else {
+			log.Debugf("process with PID %d is stopped, removing PID file", pid)
+			err = os.Remove(filepath.Join(runPath, rawPID.Name()))
+			if err != nil {
+				log.Errorf("could not remove PID file: %v", err)
+			}
 		}
 	}
+
+	return runningPIDs > 0, nil
+}
+
+// createRunDirectory creates a run directory at the specified path if it doesn't exist.
+// Note that this directory is never fully removed as the updater may
+// restart before the process is stopped.
+// Must be world writeable as we don't control who writes the PIDs
+func createRunDirectory(path string) error {
```

Comment:
Let's remove this, the initial creation needs to be done by the install script of the updater as it needs root (like the root of all packages) and the injector will create dir for the package it wants to lock if they don't exist